### ISSUE TITLE
Wrong kernel state

### ISSUE THREAD

**Issue by JohanMabille** [jupyterlab/jupyterlab#10701]
## Description

Jupyter Lab may not show the correct status of kernels that implement support for debugging, and more generally, that kernels that run the control and the shell on separate threads, as illustrated below:

![kernel_state_bug](https://user-images.githubusercontent.com/6754742/127022832-4596635a-b112-4b7f-b1ca-7419d8c07e70.gif)

## Reproduce

1. Launch two kernels with support for debugging (ipykernel, xeus-python)
2. Execute a cells that takes several seconds to complete in the first notebook
3. Switch to the second notebook before the cell has finished to execute.
4. Switch back to the first notebook, the kernel status is 'idle' while it should be 'busy'

A similar issue can be encountered while debugging: after hitting a breakpoint, any debug request that complete will result in a kernel 'idle' state whereas the code execution may nod have resumed.

## Expected behavior

JupyterLab shows the "right" kernel state.

## Proposal

There are different ways to handle this: we can see it as a JupyterLab bug, since a status message embeds a parent header which makes it possible to know to which request, and therefore to which channel, this message is associated. This would require updating JupyterLab only, more precisely [this part of the code]. 

Or wa can decide to add an extra field to the status message, the channel it is associated with. This appears to be a more robust solution to me, even if it requires more changes: JupyterLab does not have to know which channel is supposed to receive which request.

Both of these solutions fix the debugging issue (i.e. the kernel state is updated to 'idle' because of the control request). 
However, I'm not sure this would fix the kernel switching issue; either JupyterLab already stores the kernel state along with the kernel name and other kernel info somewhere. In that case, to avoid updating the kernel stated to a wrong status, JupyterLab would have to ignore all status messages associated to the debugInfo request. Or it does not store it, and we need to send a specific message to the control channel to require the status of the kernel (but this starts to become overcomplicated in my opinion).

Another solution would be to consider this as the reponsibility of the kernel: even it the kernel processes a control request, it is its responsibility to check if some code is running on the shell and fir the appropriate event.

So to summarize the different solutions:

- Jupyter Lab change only, always drop states that result from a request on the control (retrieved thanks to the parent header)
- Protocol change and JupyterLab change: always drop states that result from a request on the control (retrieved thanks to an additional field in the status message)
- Kernel change: whenever a status event is emitted, it is the responsibility of the kernel to ensure all the requests on the shell have completed.

cc @afshin @jasongrout @minrk @SylvainCorlay @jtpio @fcollonval @blink1073 as any of this change may have impacts on parts of the project that I don't know well.

**Comment by davidbrochart** [comment]
I think we should not make any change in the kernel protocol. It is always better to provide more details about what is happening in the kernel, and let the client decide what to do with it, rather than filtering the information that is given to the client.
JupyterLab shows a kernel status indicator, which indicates if the kernel is executing some code, so it is clearly a bug in JLab if this status is cleared when switching tabs, since it has all the information to know that it is not true. We could also imagine other indicators in the future, showing e.g. comm messages status separately, etc.
Thus I would advocate for the first solution, which involves changes in JupyterLab only.

**Comment by JohanMabille** [comment]
> It is always better to provide more details about what is happening in the kernel, and let the client decide what to do with it, rather than filtering the information that is given to the client.

Which is exactly what a change in the protocol would do: JupyterLab would know immediatly to which channel the status message is associated, instead of having to retrieve it via the parent header.

Changes in JupyterLab only means the part of the code I linked must be aware of all the requests and where they're supposed to be sent.

**Comment by davidbrochart** [comment]
Ah ok, I hadn't realized you wanted to add information when changing the protocol:

> Or we can decide to add an extra field to the status message, the channel it is associated with

As you said, since the client can retrieve this information through the parent header, it is not really needed, but I agree this can make it simpler for clients.

**Comment by SylvainCorlay** [comment]
Following a direct conversation on this issue with @JohanMabille on this

 - we should probably have the front-end kernel busy/idle indicator only reflect the business of the Shell channel/thread of the kernel.
 - async kernels like `akernel` can have several cells running concurrently and a single "idle" message with the parent header of an execution request only indicates the fact that one execution task was completed. Tracking the number of running tasks and properly checking the parent header in the front-end is probably the way to go.
 - if we implement the above proposal, this indicates that the "idle/busy" messages in the protocol are probably misleading.

**Comment by paravatha** [comment]
@JohanMabille, I am seeing the same issue in jupyterlab==3.1.4
is this bug introduced recently? if it is, appreciate if you could let me know which version I should roll-back to!

**Comment by fcollonval** [comment]
@paravatha you will hit this issue as soon as you use a kernel with debug capability (and JupyterLab 3).

**Comment by paravatha** [comment]
Thanks @fcollonval 
If I don't upgrade ipykernel, use the below versions..I should be good?
```
ipykernel                     5.5.5
jupyterlab                    3.1.2
```

**Comment by SylvainCorlay** [comment]
Note that this should be fixed as a byproduct of https://github.com/jupyterlab/jupyterlab/pull/10730.
